### PR TITLE
Fix nonce deserialization (Alternative Solution)

### LINKED ISSUE #629: Message `nonce` field can be Integer or String, but Kord always deserializes it as a String
```json
{"d": {"id": "990134009365401670", "tts": false, "type": 0, "flags": 0, "nonce": 6870, "author": {"id": "330770985450078208", "bot": true, "avatar": "6668e8af53a272e4cd72888d2183c97f", "username": "Aperture", "public_flags": 65536, "discriminator": "8658", "avatar_decoration": null}, "embeds": [], "member": {"deaf": false, "mute": false, "nick": null, "user": {"id": "330770985450078208", "bot": true, "avatar": "6668e8af53a272e4cd72888d2183c97f", "username": "Aperture", "public_flags": 65536, "discriminator": "8658", "avatar_decoration": null}, "flags": 0, "roles": ["753439237856165889", "751233962428596274"], "avatar": null, "pending": false, "joined_at": "2020-09-10T02:18:32.579000+00:00", "premium_since": null, "communication_disabled_until": null}, "pinned": false, "content": "`[02:59:03]` :outbox_tray: RGN-Rafaelk#​4169 (`568682176087064576`) left the server", "guild_id": "751223482515325088", "mentions": [], "timestamp": "2022-06-25T05:59:03.633000+00:00", "channel_id": "752735099971436584", "components": [], "attachments": [], "mention_roles": [], "edited_timestamp": null, "mention_everyone": false, "referenced_message": null}, "s": 246086, "t": "MESSAGE_CREATE", "op": 0}
```

```
kotlinx.serialization.json.internal.JsonDecodingException: Unexpected JSON token at offset 113: Expected quotation mark '"', but had '0' instead at path: $.d.nonce
```

### PR DESCRIPTION
**Pull request by MrPowerGamerBR**
image

Fixes #629

Only #632 or #630 must be merged, they both fix the same issue, but in different ways.

### CODE REVIEW

File: common/src/main/kotlin/entity/DiscordMessage.kt [hunk 1]
```diff
@@ -114,6 +121,26 @@ public data class DiscordMessage(
     val thread: Optional<DiscordChannel> = Optional.Missing()
 )
 
+internal object NonceAsStringSerializer : KSerializer<String> {
```
Comment (Lukellmann) [inline common/src/main/kotlin/entity/DiscordMessage.kt:124]: I've implemented something similar [here](https://github.com/kordlib/kord/blob/0.8.x/common/src/main/kotlin/entity/AuditLog.kt#L319). I suggest a refactor to use the same serializer in both places.
A good place to put this serializer would be [here](https://github.com/kordlib/kord/tree/0.8.x/common/src/main/kotlin/serialization).

**Review by Lukellmann — CHANGES_REQUESTED** [review]
I would also go for this solution, this doesn't introduce yet another type.

File: common/src/main/kotlin/entity/AuditLog.kt [hunk 2]
```diff
@@ -310,39 +309,11 @@ public sealed class AuditLogChangeKey<T>(public val name: String, public val ser
...
-                val jsonPrimitive = value.toIntOrNull()?.let { JsonPrimitive(it) } ?: JsonPrimitive(value)
-                encoder.encodeSerializableValue(backingSerializer, jsonPrimitive)
-            }
-
-            override fun deserialize(decoder: Decoder): String {
-                val jsonPrimitive = decoder.decodeSerializableValue(backingSerializer)
-                return if (jsonPrimitive.isString) jsonPrimitive.content else jsonPrimitive.int.toString()
-            }
-        }
     }
```
Comment (Lukellmann) [inline common/src/main/kotlin/entity/AuditLog.kt:317]:
```suggestion
```
the block is no longer needed

File: common/src/main/kotlin/entity/DiscordMessage.kt [hunk 3]
```diff
@@ -13,6 +14,13 @@ import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
 import kotlinx.serialization.descriptors.SerialDescriptor
 import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.JsonContentPolymorphicSerializer
+import kotlinx.serialization.json.JsonDecoder
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.int
+import kotlinx.serialization.json.intOrNull
+import kotlinx.serialization.json.jsonObject
+import kotlinx.serialization.json.jsonPrimitive
```
Comment (Lukellmann) [inline common/src/main/kotlin/entity/DiscordMessage.kt:23]:
```suggestion
```
we no longer need these imports

**Review by Lukellmann — COMMENTED** [review]
looks good overall, just some minor suggestions

**Comment by MrPowerGamerBR** [comment]
For some reason the "Add suggestion to batch" was grayed out, that's why I needed to apply all suggestions one by one 😭

**Comment by Lukellmann** [comment]
> For some reason the "Add suggestion to batch" was grayed out, that's why I needed to apply all suggestions one by one 😭

iirc applying batch suggestions is only available in the "Files changed" tab

**Pull request merged**